activation/validation.go

```diff
@@ -264,3 +290,140 @@ func (v *Validator) PositioningAtx(
 	}
 	return nil
 }
+
+type verifyChainOpts struct {
+	assumedValidTime time.Time
+	trustedNodeID    types.NodeID
+}
+
+type verifyChainOption func(*verifyChainOpts)
+
+// assumeValidBefore configures the validator to assume that ATXs received before the given time are valid.
+func assumeValidBefore(val time.Time) verifyChainOption {
+	return func(o *verifyChainOpts) {
+		o.assumedValidTime = val
+	}
+}
+
+// withTrustedID configures the validator to assume that ATXs created by the given node ID are valid.
+func withTrustedID(val types.NodeID) verifyChainOption {
+	return func(o *verifyChainOpts) {
+		o.trustedNodeID = val
+	}
+}
+
+var ErrInvalidChain = errors.New("invalid ATX chain")
+
+func verifyChain(
+	ctx context.Context,
+	db sql.Executor,
+	id, goldenATXID types.ATXID,
+	validator nipostValidator,
+	log *zap.Logger,
+	opts ...verifyChainOption,
+) error {
+	log.Info("verifying ATX chain", zap.Stringer("atx_id", id))
+	options := verifyChainOpts{}
+	for _, opt := range opts {
+		opt(&options)
+	}
+	return verifyChainWithOpts(ctx, db, id, goldenATXID, validator, log, options)
+}
+
+func verifyChainWithOpts(
+	ctx context.Context,
+	db sql.Executor,
+	id, goldenATXID types.ATXID,
+	validator nipostValidator,
+	log *zap.Logger,
+	opts verifyChainOpts,
+) error {
+	atx, err := atxs.Get(db, id)
+	if err != nil {
+		return fmt.Errorf("get atx: %w", err)
+	}
+
+	switch {
+	case atx.Validity() == types.Valid:
+		log.Debug("not verifying ATX chain", zap.Stringer("atx_id", id), zap.String("reason", "already verified"))
+		return nil
+	case atx.Validity() == types.Invalid:
+		log.Debug("not verifying ATX chain", zap.Stringer("atx_id", id), zap.String("reason", "invalid"))
+		return errors.Join(ErrInvalidChain, errors.New("atx is marked as invalid"))
```

Comment:
With a custom type this would just become:

```suggestion
		return &ErrInvalidChain{ID: atx.ID()}
```

Reply:
I used errors.Join here to differentiate between "atx is already marked invalid in DB" vs "I checked this ATX post and it is invalid". Probably not needed though :shrug: .